### PR TITLE
[optimizer] Move optimizer out of layer and optimizer tensors to their weights

### LINKED ISSUE #630: Copy for layers is broken
Copy member function for layers aims to create a copy of the current layer.
However, for the optimizer, it just uses copy assignment operation which does not deep copy the tensors inside the optimizer.

### PR DESCRIPTION
**Pull request by kparichay**
**Commit 1**: [optimizer] Move optimizer variables to weights
    
Move optimizer variables to weights
    Now all the weight related tensors are handled by weights themselves
    So, optimizer can be shared across all layers, no need to create new
    copies for all layers
    
**Commit 2**: [optimizer] Simplify optimizer initialize
    
As there is just one optimizer and shared by layers, it must be initialized just once by the neural network.
    Also, addOptimizerVariables() moved out separately from initialize() as initialize() should work
    on optimizers parameters and should not need list of weights.
    
Also remove set_tensor argument which was redundant

**Commit 3**: [optimizer] Move optimizer out of layer
    
This patch moves optimizer out of layer.
    Now backwarding just calculates derivatives and gradient
    but does not applies the gradient.
    This gradient applying is done by the model.
    
Layer still support applyGradient operation but requires optimizer
    as an argument.
    This decouples layers from optimizers and can operate independently.

**Commit 4**: [network] Rework the backwarding
    
- remove forwarding from backwarding
    backwarding should just do backwarding and no more
- moved backwarding back to neuralnetwork so that graph
    does not has to care about how to backward etc.
    Graph just provides iterators for iterating the graph
    in reverse. Graph does not know that layers have backwarding etc.
    
Also this removes dependency of graph from optimizer.

See Also #774
Resolves #630

**Self evaluation:**
    1. Build test: [x]Passed [ ]Failed [ ]Skipped
    2. Run test: [x]Passed [ ]Failed [ ]Skipped
    
Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/models/neuralnet.cpp [hunk 1]
```diff
@@ -152,6 +152,13 @@ int NeuralNetwork::compile() {
 
   model_graph.topologicalSort();
 
+  auto sorted = model_graph.getSorted();
```
Comment (zhoonit) [inline nntrainer/models/neuralnet.cpp:155]: auto & here?

File: nntrainer/graph/network_graph.h [hunk 2]
```diff
@@ -191,19 +191,26 @@ class NetworkGraph {
...
 
   /**
-   * @brief     backwarding network graph
-   * @param[in] input data
-   * @param[in] iteration
+   * @brief     getter of ordered graph
+   * @retval    ordered LayerNode list
    */
-  void backwarding(sharedConstTensors input, int iteration);
+  std::vector<LayerNode> getSorted() { return Sorted; }
```
Comment (zhoonit) [inline nntrainer/graph/network_graph.h:197]: `const std::vector<LayerNode> &` to prevent copy but maintain value semantics?

**Review by zhoonit — APPROVED** [review]
LGTM with insignificant comments :) Nice work done.

File: nntrainer/layers/layer.cpp [hunk 3]
```diff
@@ -82,8 +76,6 @@ void Layer::copy(std::shared_ptr<Layer> l) {
   for (auto const &w : weights)
     weights.push_back(w.clone());
 
-  // TODO: fix this #630
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/layers/layer.cpp:84]: Its already there. You mean add to commit description?

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 4]
```diff
@@ -159,14 +159,18 @@ class nntrainer_abstractLayer : public ::testing::Test {
...
+    status = opt->setProperty(input_str);
     EXPECT_EQ(status, ML_ERROR_NONE);
-    status = layer.setOptimizer(op);
+
+    status = opt->initialize();
     EXPECT_EQ(status, ML_ERROR_NONE);
 
+    EXPECT_NO_THROW(opt->addOptimizerVariable(layer.getWeightsRef()));
+    // status = layer.setOptimizer(op);
+    // EXPECT_EQ(status, ML_ERROR_NONE);
```
[the remark this answers is not in the document]
Comment (kparichay) [reply test/unittest/unittest_nntrainer_layers.cpp:172]: removed.

File: nntrainer/layers/layer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhoonit) [reply nntrainer/layers/layer.cpp:84]: My bad, I missed it 👍

File: nntrainer/models/neuralnet.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/models/neuralnet.cpp:155]: done.

File: nntrainer/graph/network_graph.h (hunk 2 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.h:197]: done.

File: nntrainer/layers/layer_internal.h [hunk 5]
```diff
@@ -134,22 +134,26 @@ class Layer : public ml::train::Layer {
    * @brief     Apply the gradient for the layer
    * @param[in] iteration Iteration value for the Optimizer
    */
-  virtual void applyGradient(unsigned int iteration);
+  virtual void applyGradient(unsigned int iteration,
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/layers/layer_internal.h:137]: Done.

File: nntrainer/tensor/weight.h [hunk 6]
```diff
@@ -150,14 +150,33 @@ class Weight : public Var_Grad {
...
     Var_Grad::reset(dim, train);
   }
 
+  /**
+   * @brief Add optimizer variables
+   * @param dim Optimizer variable dimension
+   */
+  void addOptimizerVariable(const TensorDim &dim) {
+    opt_vars.emplace_back(dim);
+    opt_vars.back().setZero();
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/tensor/weight.h:164]: It might be common but cant guarantee.
Added a TODO.

**Review by jijoongmoon — APPROVED** [review]
LGTM
